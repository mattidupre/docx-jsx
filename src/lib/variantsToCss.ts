import type { CSSProperties } from 'react';
import { kebabCase } from 'lodash';
import {
  TYPOGRAPHY_CSS_KEYS,
  INTRINSIC_TYPOGRAPHY_OPTIONS,
  type Variants,
  INTRINSIC_VARIANT_TAG_NAMES,
} from '../entities/typography';
import type { PrefixesConfig } from '../entities';
import { getValueOf } from '../utils/object';

const TYPOGRAPHY_CSS_PROPERTIES = TYPOGRAPHY_CSS_KEYS.map(kebabCase);

type CssRule = [string, CSSProperties];

type CssRules = Array<CssRule>;

/**
 * Convert a variant value to a valid CSS property.
 * This allows custom variables to be passed into variants as arrays.
 */
export const toCssValue = (
  value: unknown,
  options: { prefixes: PrefixesConfig },
): undefined | string => {
  if (!value || (Array.isArray(value) && !value.length)) {
    return undefined;
  }
  const [currentValue, ...remainingValues] = [value].flat();

  if (!currentValue) {
    return toCssValue(remainingValues, options);
  }

  const isCurrentValueCssVariable =
    typeof currentValue === 'string' && currentValue.startsWith('--');

  if (remainingValues.length) {
    if (!isCurrentValueCssVariable) {
      throw new TypeError('Only the last value may be a CSS variable');
    }
    return `var(${currentValue}, ${toCssValue(remainingValues, options)})`;
  }

  if (isCurrentValueCssVariable) {
    throw new TypeError('The last value must not be a CSS variable');
  }

  return String(currentValue);
};

export const variantsToCssRules = (
  variants: Variants,
  options: {
    prefixes: Pick<PrefixesConfig, 'cssVariable' | 'variantClassName'>;
  },
): CssRules => {
  const { prefixes } = options;

  // TODO: How to handle highlight, <sup>, <sub>?
  const rules: CssRules = [];

  rules.push([
    '*',
    TYPOGRAPHY_CSS_PROPERTIES.reduce(
      (declarations, property) => ({
        ...declarations,
        [property]: `var(--${prefixes.cssVariable}-${kebabCase(property)})`,
      }),
      {} as CSSProperties,
    ),
  ]);

  // Override vars are defined inline in the HTML.

  for (const tagName in INTRINSIC_TYPOGRAPHY_OPTIONS) {
    rules.push([
      `:where(${tagName})`,
      Object.fromEntries(
        Object.entries(
          INTRINSIC_TYPOGRAPHY_OPTIONS[
            tagName as keyof typeof INTRINSIC_TYPOGRAPHY_OPTIONS
          ],
        ).map(([propertyName, propertyValue]) => {
          if (
            !TYPOGRAPHY_CSS_KEYS.includes(
              propertyName as (typeof TYPOGRAPHY_CSS_KEYS)[number],
            )
          ) {
            return [];
          }
          return [
            [`--${prefixes.cssVariable}-${kebabCase(propertyName)}`],
            toCssValue(propertyValue, options),
          ];
        }),
      ),
    ]);
  }

  // Variant vars are attached to variant class names .[prefix]-[variant-name]
  for (const variantName in variants) {
    const variantKebab = kebabCase(variantName);

    const variantSelector = `.${prefixes.variantClassName}-${kebabCase(
      variantName,
    )}`;
    const tagNameSelector = getValueOf(
      variantName,
      INTRINSIC_VARIANT_TAG_NAMES,
    );

    const selector = tagNameSelector
      ? `:where(${tagNameSelector}, ${variantSelector})`
      : variantSelector;

    rules.push([
      // Variant tags set variant CSS vars.
      selector,
      Object.fromEntries(
        Object.entries(variants[variantName]).map(
          ([propertyName, propertyValue]) => {
            return [
              [
                `--${prefixes.cssVariable}-${variantKebab}-${kebabCase(
                  propertyName,
                )}`,
              ],
              toCssValue(propertyValue, options),
            ];
          },
        ),
      ),
    ]);

    rules.push([
      // Properties in variant tags default to override CSS vars but fall back to variant CSS vars.
      `:where(${selector} *)`,
      TYPOGRAPHY_CSS_PROPERTIES.reduce(
        (declarations, property) => ({
          ...declarations,
          [property]: `var(--${prefixes.cssVariable}-${property}, var(--${prefixes.cssVariable}-${variantKebab}-${property}))`,
        }),
        {} as CSSProperties,
      ),
    ]);
  }

  return rules;
};
