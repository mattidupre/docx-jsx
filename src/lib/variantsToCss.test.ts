import { expect, test, describe } from 'vitest';
import {
  createMockPrefixesConfig,
  createMockVariantsConfig,
} from '../fixtures';
import { variantsToCssRules } from './variantsToCss';

describe('joinNestedKeys', () => {
  test('passing deep object', () => {
    const MOCK_VARIANTS = createMockVariantsConfig();
    const MOCK_PREFIXES = createMockPrefixesConfig();

    const cssRules = variantsToCssRules(MOCK_VARIANTS, {
      prefixes: MOCK_PREFIXES,
    });
    for (const rule of cssRules) {
      console.log(`${rule[0]}: \n${JSON.stringify(rule[1], undefined, 2)}\n`);
    }
  });
});
