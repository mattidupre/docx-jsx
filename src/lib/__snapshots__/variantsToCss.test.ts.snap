// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`joinNestedKeys > passing deep object 1`] = `
[
  [
    "*",
    {
      "break-inside": "var(--mock-variable-break-inside)",
      "color": "var(--mock-variable-color)",
      "font-family": "var(--mock-variable-font-family)",
      "font-size": "var(--mock-variable-font-size)",
      "font-style": "var(--mock-variable-font-style)",
      "font-weight": "var(--mock-variable-font-weight)",
      "line-height": "var(--mock-variable-line-height)",
      "text-align": "var(--mock-variable-text-align)",
      "text-decoration": "var(--mock-variable-text-decoration)",
      "text-transform": "var(--mock-variable-text-transform)",
    },
  ],
  [
    ":where(b)",
    {
      "--mock-variable-font-weight": "bold",
    },
  ],
  [
    ":where(strong)",
    {
      "--mock-variable-font-weight": "bold",
    },
  ],
  [
    ":where(em)",
    {
      "--mock-variable-font-style": "italic",
    },
  ],
  [
    ":where(u)",
    {
      "--mock-variable-text-decoration": "underline",
    },
  ],
  [
    ":where(s)",
    {
      "--mock-variable-text-decoration": "line-through",
    },
  ],
  [
    ":where(sup)",
    {},
  ],
  [
    ":where(sub)",
    {},
  ],
  [
    ".mock-class-name-mock-variant",
    {
      "--mock-variable-mock-variant-color": "var(--mock-color, #00ff00)",
      "--mock-variable-mock-variant-font-weight": "bold",
    },
  ],
  [
    ":where(.mock-class-name-mock-variant *)",
    {
      "break-inside": "var(--mock-variable-break-inside, var(--mock-variable-mock-variant-break-inside))",
      "color": "var(--mock-variable-color, var(--mock-variable-mock-variant-color))",
      "font-family": "var(--mock-variable-font-family, var(--mock-variable-mock-variant-font-family))",
      "font-size": "var(--mock-variable-font-size, var(--mock-variable-mock-variant-font-size))",
      "font-style": "var(--mock-variable-font-style, var(--mock-variable-mock-variant-font-style))",
      "font-weight": "var(--mock-variable-font-weight, var(--mock-variable-mock-variant-font-weight))",
      "line-height": "var(--mock-variable-line-height, var(--mock-variable-mock-variant-line-height))",
      "text-align": "var(--mock-variable-text-align, var(--mock-variable-mock-variant-text-align))",
      "text-decoration": "var(--mock-variable-text-decoration, var(--mock-variable-mock-variant-text-decoration))",
      "text-transform": "var(--mock-variable-text-transform, var(--mock-variable-mock-variant-text-transform))",
    },
  ],
  [
    ":where(h1, .mock-class-name-title)",
    {
      "--mock-variable-title-color": "#0000ff",
    },
  ],
  [
    ":where(:where(h1, .mock-class-name-title) *)",
    {
      "break-inside": "var(--mock-variable-break-inside, var(--mock-variable-title-break-inside))",
      "color": "var(--mock-variable-color, var(--mock-variable-title-color))",
      "font-family": "var(--mock-variable-font-family, var(--mock-variable-title-font-family))",
      "font-size": "var(--mock-variable-font-size, var(--mock-variable-title-font-size))",
      "font-style": "var(--mock-variable-font-style, var(--mock-variable-title-font-style))",
      "font-weight": "var(--mock-variable-font-weight, var(--mock-variable-title-font-weight))",
      "line-height": "var(--mock-variable-line-height, var(--mock-variable-title-line-height))",
      "text-align": "var(--mock-variable-text-align, var(--mock-variable-title-text-align))",
      "text-decoration": "var(--mock-variable-text-decoration, var(--mock-variable-title-text-decoration))",
      "text-transform": "var(--mock-variable-text-transform, var(--mock-variable-title-text-transform))",
    },
  ],
]
`;
